src/output-json-anomaly.c

```diff
@@ -0,0 +1,384 @@
+ *
+ * Logs anomalies in JSON format.
+ *
+ */
+
+#include "suricata-common.h"
+#include "debug.h"
+#include "detect.h"
+#include "flow.h"
+#include "conf.h"
+
+#include "threads.h"
+#include "tm-threads.h"
+#include "threadvars.h"
+#include "util-debug.h"
+
+#include "util-misc.h"
+
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "util-logopenfile.h"
+
+#include "output.h"
+#include "output-json.h"
+#include "output-json-anomaly.h"
+
+#include "util-byte.h"
+#include "util-privs.h"
+#include "util-print.h"
+#include "util-proto-name.h"
+#include "util-optimize.h"
+#include "util-buffer.h"
+#include "util-validate.h"
+
+#define MODULE_NAME "JsonAnomalyLog"
+
+#ifdef HAVE_LIBJANSSON
+
+#define JSON_STREAM_BUFFER_SIZE 4096
+
+typedef struct AnomalyJsonOutputCtx_ {
+    LogFileCtx* file_ctx;
+    uint16_t flags;
+    uint32_t payload_buffer_size;
+    OutputJsonCommonSettings cfg;
+} AnomalyJsonOutputCtx;
+
+typedef struct JsonAnomalyLogThread_ {
+    /** LogFileCtx has the pointer to the file and a mutex to allow multithreading */
+    LogFileCtx* file_ctx;
+    MemBuffer *json_buffer;
+    MemBuffer *payload_buffer;
+    AnomalyJsonOutputCtx* json_output_ctx;
+} JsonAnomalyLogThread;
+
+static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *p)
+{
+    uint8_t i;
+    json_t *js;
+
+    if (p->events.cnt == 0) {
+        return TM_ECODE_OK;
+    }
+    AnomalyJsonOutputCtx *json_output_ctx = aft->json_output_ctx;
+
+
+    for (i = 0; i < p->events.cnt; i++) {
+        MemBufferReset(aft->json_buffer);
+
+        js = CreateJSONHeader(p, LOG_DIR_PACKET, "anomaly");
+        if (unlikely(js == NULL))
+            return TM_ECODE_OK;
+
+        JsonAddCommonOptions(&json_output_ctx->cfg, p, p->flow, js);
+
+        char buf[(32 * 3) + 1];
+        PrintRawLineHexBuf(buf, sizeof(buf), GET_PKT_DATA(p), GET_PKT_LEN(p) < 32 ? GET_PKT_LEN(p) : 32);
+
+        json_t *ajs = json_object();
+        if (unlikely(ajs == NULL)) {
+            json_decref(js);
+            return TM_ECODE_OK;
+        }
+
+        JsonFiveTuple((const Packet *)p, LOG_DIR_PACKET, js);
+
+        JsonAddCommonOptions(&json_output_ctx->cfg, p, p->flow, js);
+
+        if (p->tenant_id > 0)
+            json_object_set_new(ajs, "tenant_id", json_integer(p->tenant_id));
+
+        uint8_t event_code = p->events.events[i];
+        const char *event;
+        if (EVENT_IS_DECODER_PACKET_ERROR(event_code)) {
+            event = DEvents[event_code].event_name;
+        } else {
+            event = (const char *) "Unknown";
+        }
+        json_object_set_new(ajs, "event", json_string(event));
+
+        /* anomaly */
+        json_object_set_new(js, "anomaly", ajs);
+        OutputJSONBuffer(js, aft->file_ctx, &aft->json_buffer);
+
+        json_object_clear(js);
+        json_decref(js);
+    }
+
+    return TM_ECODE_OK;
+}
+
+static int AnomalyJsonDecoderEvent(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *p)
```

Comment:
given the simple logic here this can probably be merged with AnomalyJson

Reply:
Good idea.